File: backend/internal/domain/analytics.go
package domain

import "errors"

type PageView struct {
	PageLocation string `json:"location"`
	PageTitle    string `json:"title"`
}

func (p PageView) Validate() error {
	if p.PageLocation == "" {
		return errors.New("pageLocation missing")
	}
	if p.PageTitle == "" {
		return errors.New("pageTitle missing")
	}

	return nil
}

File: apps/frontend/src/lib/services/analytics.ts
import { APIRoutes } from '@/routes/api-routes';

export const AnalyticsService = {
  pageView: async ({
    location,
    title,
  }: {
    location: string;
    title: string;
  }) => {
    try {
      const res = await fetch(APIRoutes.pageView, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          location,
          title,
        }),
      });

      if (!res.ok) {
        throw new Error(
          `failed to page view (status: ${res.status} - ${res.statusText})`,
        );
      }

      return;
    } catch (error) {
      console.error('AnalyticsService.pageView error: ', error);
      return;
    }
  },
};
